sceptre/hooks/__init__.py

```diff
@@ -28,6 +39,42 @@ def run(self):
         pass  # pragma: no cover
 
 
+class HookProperty(object):
+    """
+    This is a descriptor class used to store an attribute that may contain
+    Hook objects. Used to setup Hooks when added as a attribute. Supports
+    nested dictionary and lists.
+
+    :param name: Attribute suffix used to store the property in the instance.
+    :type name: string
+    """
+    def __init__(self, name):
+        self.name = "_" + name
+        self.logger = logging.getLogger(__name__)
+
+    def __get__(self, instance, type):
+        """
+        Attribute getter for Hook containing data structure.
+
+        :return: The attribute stored with the suffix ``name`` in the instance.
+        :rtype: dict or list
+        """
+        return getattr(instance, self.name)
+
+    def __set__(self, instance, value):
+        """
+        Attribute setter which adds a stack reference to any hooks in the
+        data structure `value` and calls the setup method.
+
```

Comment:
I think documenting the types of instance and value would be useful; even though they can be ascertained by reading the rest of the code, it would make it easier to understand what's going on quickly.